### ISSUE TITLE
Forms and Fields Architechture

### ISSUE THREAD

**Issue by HipsterBrown** [namely/styleguide#214]
Forms are a major part of how most people interact with Namely, and the experience across HCM is inconsistent to say the least. 

We have the latest designs in the [styleguide doc](https://basecamp.com/1899881/projects/666779/messages/29681611?enlarge=191900557#attachment_191900557) and most of the [base styles & CSS utilities](https://github.com/namely/styleguide/blob/develop/src/scss/elements/_forms.scss) in place. We event have some [basic form components](https://github.com/namely/styleguide/tree/develop/src/js/components/forms/fields). This RFC is for us to settle on a standard for constructing forms and field components to be used across HCM and other Namely applications. 

Things to consider: 
- Do we keep the ["Form App" pattern](https://github.com/namely/namely/blob/develop/app/assets/javascripts/ui/onboarding/templates/form_app.es6)?
  - If no, how do we replace it for the better?
- How are Fields different from plain form elements?
- Do we replicate the current [Field component](https://github.com/namely/namely/blob/develop/app/assets/javascripts/ui/fields/field.coffee) functionality?
- How does we consistently get values from Fields?
- What default Fields do we need available for HCM, i.e. Date, SSN, Image?
- How do we validate forms and fields?
- How do we serialize and send form data?

Please comment with any resources or references you find helpful for supporting a certain conventions. The more context behind a decision, the easier it will be to come to a shared standard. Places like [KhanAcademy](https://github.com/Khan?utf8=%E2%9C%93&query=React) and [AirBnb](https://github.com/airbnb/javascript/tree/master/react) might be great inspirations for this work since they often share how they solve similar problems.

**Comment by loudhearted** [comment]
I'm not a fan of the current Field component setup; it looks like it's main job is to render a more specific type of field component inside it based on a switch statement, and the individual specific fields don't seem to share all that much in this Field parent component, so why not just create fully separate DateField, TextField etc. components?

Unless I'm totally missing something, these fields appear to have more differences than shared code, and thus forces it to hold onto props that not all of its potential children need, like `dateFormat`, or else perform even more goofy if/switch statements to handle certain child fields, like [this](https://github.com/namely/namely/blob/develop/app/assets/javascripts/ui/fields/field.coffee#L46). It also makes, and would make, extensive customization of individual field types tougher than it needs to be.

Also, the current setup makes a group of fields look like this, which drives me nuts:

```js
<FieldGroup>
<Field>
<Field>
<Field>
</FieldGroup>
```
whereas this would be faster to parse:
```js
<FieldGroup>
<TextField>
<DateField>
<SsnField>
</FieldGroup>
```

In short, I'm for field type self-rule.

**Comment by HipsterBrown** [comment]
@loudhearted I believe the original intent behind the Field component was to be able to take a collection of field types from the Namely API, then render a form by iterating over that collection and passing in a type into the Field component to get the right component back. 

I agree it's been abused a bit, like you were referencing. That is something we should document: when should & shouldn't the Field component be used.

**Comment by loudhearted** [comment]
@HipsterBrown Maybe we could rename `Field` to `DynamicField`, or even something goofy but more illustrative like `ShapeshiftField` or `MorphingField`, to denote that it is only the empty husk of a field that should only be used to turn into another field type. 

I'm also curious if there'd be a better way altogether to render a list of fields from the API to eliminate this altogether, but no solid ideas at the moment.

I'd also be in favor of or `MightyMorphinPowerField`

**Comment by HipsterBrown** [comment]
As much as I like the idea of a component called `MightyMorphinPowerField`, I think `DynamicField` is a good name for describing the functionality of this component and what it's used for.

**Comment by keithpops** [comment]
How about a field factory?

https://medium.com/@chekofif/how-to-grok-higher-order-classes-69452ba6d095#.it54oago8

**Comment by keithpops** [comment]
Also, instead of `FieldGroup` would it make sense to call them `Well`s or something similar? The idea being is that our groups sometimes do not always include fields. I'm borrowing this idea from: http://getbootstrap.com/components/#wells

**Comment by keithpops** [comment]
Another brain dump:

I'd love to see this for a `Form` component:
```
<AltContainer flux={FormFlux}>
  <Form autosave={true} endpoint={'/endpoint'}>

    <Well>
      <Field.String label="Name" errors={[ 'an', 'array', 'of', 'problems', 'via', 'state' ]} />
      <Field.Address label="Address" />
    </Well>
    <Field.Currency label="Salary" />

    <ButtonGroup>
      <Button label="Save" onClick={this.props.flux.actions.FormActions.save} />
    </ButtonGroup>

  </form>
</AltContainer>
```

Perhaps also worth a look into https://github.com/mozilla/localForage to save the form data locally (and across browser tabs) which could be handled with a `Form` flux pattern.

**Comment by HipsterBrown** [comment]
@keithpops 
> would it make sense to call them Wells or something similar?

The Wells, as Bootstrap describes them, look like more of a prescribed style than any sort of useful grouping component. What does a component like a Well give over using a native `fieldset` element?

> The idea being is that our groups sometimes do not always include fields.

Could you clarify this a bit more? What else would be in a form that's not a field?

> I'd love to see some variation of this for a Form component

I was imagining a similar architecture for the base form component as well. The Form should have an exposed `value` property that returns the serialized values of all child fields. I also agree getting some local persistence in the data layer would be rad.
